### ISSUE TITLE
Tracker and guide: Update to Godot 4.3

### ISSUE THREAD

**Issue by NathanLovato** [gdquest-demos/godot-shaders#53]
This issue focuses on porting the Godot Shaders repository from Godot 3 to Godot 4.3. The repository contains a collection of shaders and demo scenes for 2D and 3D games. The goal is to update the shaders and demo scenes to work with Godot 4.3, so that people can use them in their projects.

## How to contribute

I saved a copy of the current shaders for Godot 3 in a branch. We can work directly on the `main` branch to port the project to Godot 4.3. Here's how you can help.

Every file needs to be checked and compared against the Godot 3 version to make sure it works correctly in Godot 4.3. The scenes, shaders, and GDScript files generally need to be updated to use the new features and syntax of the engine.

Below you will find a list of files that have not been updated yet. I will update the ones left to do as new files get updated and work in Godot 4.3.

### Claiming files to update

If you want to help, please:

- Check the files other people have claimed in the comments below.
- List the files you'll work on in a new comment.

I'll update the list periodically as you claim some. This is to avoid having multiple people updating the same files in parallel.

### Updating the files

To update the files, you'll need to:

1. Fork the repository and create a new branch.
2. Update the files to work with Godot 4.3. A handful of shaders will require good familiarity with shader programming to update. Most importantly the 3D water and the energy field ones. Others just need to be tested and updated to the new Godot 4 API. You can use this guide for a list of changes: [Godot 4.3 shader API changes](https://docs.godotengine.org/en/stable/tutorials/migrating/upgrading_to_godot_4.html#updating-shaders). The [Godot 4.3 shading language documentation](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/shading_language.html#doc-shading-language) offers a reference for the shading language.
3. Test the demo scene associated with a given shaders to make sure it works correctly.
4. Create a pull request to merge your changes back into the main repository.

Here's a video running through the process of tracing where a shader is used and making some quick adjustments to it: [Godot 4.3: How to update shaders](https://vimeo.com/1009590873/73ebfa916d?share=copy).

## Shaders and dependencies to update

Each checkbox represents a shader. If it's unchecked, it means that this shader and its dependencies haven't been updated yet. If you want to work on some shaders, please write a comment below listing the files you'll update.

**Note:** Be sure to check comments below for files others have claimed. I will not be able to update the list in real-time, so please coordinate with others to avoid working on the same files.

- [ ] advanced_toon.gdshader
- [ ] baked_sprite_glow.gdshader
- [ ] clouds2D.gdshader
- [ ] compose.gdshader
- [ ] dissolve2D.gdshader
- [ ] dissolve2D_mask.gdshader
- [ ] dissolve.gdshader
- [ ] flag_3d.gdshader
- [ ] force_field.gdshader
- [ ] fresnel_crystal.gdshader
- [ ] gaussian_blur.gdshader
- [ ] gaussian_blur_optimized.gdshader
- [ ] glow_prepass.gdshader
- [ ] interactive_snow.gdshader
- [ ] invert.gdshader
- [ ] matcap.gdshader
- [ ] outline2D_inner.gdshader
- [ ] outline2D_inner_outer.gdshader
- [ ] outline2D_outer.gdshader
- [ ] outline3D.gdshader
- [ ] outline3D_smooth_normals_color.gdshader
- [ ] PalettSwap2D.gdshader
- [ ] particle_bridge_particle.gdshader
- [ ] particle_bridge_spatial.gdshader
- [ ] perlin_noise.gdshader
- [ ] pixel_perfect_outline3D.gdshader
- [ ] pointilism.gdshader
- [ ] random_noise.gdshader
- [ ] rim_control.gdshader
- [ ] shockwave_3d.gdshader
- [ ] shockwave.gdshader
- [ ] SpecularControl.gdshader
- [ ] sphere_mask.gdshader
- [ ] stencil_mask.gdshader
- [ ] stylized_fire.gdshader
- [ ] stylized_liquid.gdshader
- [ ] stylized_waterfall.gdshader
- [ ] texture_mix.gdshader
- [ ] unlit_directional_tint.gdshader
- [ ] value_noise.gdshader
- [ ] voronoi_noise.gdshader
- [ ] water_2D.gdshader
- [ ] water_3d.gdshader
- [ ] wind_grass.gdshader
- [ ] wind_texture.gdshader
- [ ] wind_uv.gdshader
- [ ] xray_2d_mask.gdshader
- [ ] xray_3d_mask.gdshader
- [ ] VisualShader/shockwave.tres

**Comment by tetrapod00** [comment]
Claiming:
- [ ]  flag_3d.gdshader

**Comment by Leonardo-Rocha** [comment]
Claiming:
- [ ] texture_mix.gdshader
- [ ] (not listed) fixing `Demo3DEnvironment.tscn`

**Comment by Excoh** [comment]
Claiming:
- [ ] dissolve2D.gdshader
- [ ] dissolve2D_mask.gdshader
- [ ] dissolve.gdshader

**Comment by jrassa** [comment]
Claiming:
* [ ] Restore visuals in DemoPickerUI.tscn: UI theme resource and background image

**Comment by tetrapod00** [comment]
I don't have time to investigate right now, but I'd consider removing or replacing the toon shader currently included. The current shader uses viewports and has a visible frame of lag in the editor. In 4.x, there are approaches that work entirely within the shader and shader includes, without the need for viewports or a setup script in the scene. 

One example of a cel shader that works without viewports is https://godotshaders.com/shader/complete-cel-shader-for-godot-4/ (https://github.com/eldskald/godot4-cel-shader), there are probably others and perhaps one has a compatible license and a creator who wants it included in this repo.

**Comment by tetrapod00** [comment]
The water shader can also use some improvements, but that one is much more suited to being upgraded in place. I'm willing and able to make the water shader changes.

**Comment by albertok** [comment]
Claiming stylized_fire - #63

**Comment by nan0m** [comment]
claiming stylized_liquid -> also finished at [#65 ](https://github.com/gdquest-demos/godot-shaders/pull/65)

**Comment by NathanLovato** [comment]
@tetrapod00 regarding the advanced toon shader, it is an example of emulating deferred rendering in Godot. This approach is used to capture different light information in separate rendering channels and be able to control the effect of various lights on different models at the material level.

The shader could probably be adapted to work with forward rendering as much of the logic is just common rendering techniques: anisotropy, metallic, etc.

**Comment by tetrapod00** [comment]
I personally would not expect an example on how to emulate deferred rendering from this set of example shaders. Like you said, it's slightly working against the engine. I *might* expect some example of how to render out to a "custom buffer", using a technique something like https://github.com/thompsop1sou/custom-screen-buffers. If there is such an example, ideally it would be separated from the toon shader example, since a good toon shader is possible without emulating deferred or using viewports. I still haven't looked closer at the existing toon shader though.

**Comment by arukagiri** [comment]
Hi!
I'm trying to learn shaders at the moment so if there is no rush with these I'd like to claim some for practice!

Claiming:

- [ ] clouds2D.gdshader
- [ ] glow_prepass.gdshader
- [ ] wind_grass.gdshader

**Comment by NathanLovato** [comment]
@arukagiri Thank you very much for your help! I have a heavy workload right now but wanted to let you know I saw your PRs and will get to merging them as soon as possible.

**Comment by macodeth** [comment]
Hi, I'm claiming: 
- [x] outline2D_inner.gdshader
- [x] outline2D_inner_outer.gdshader
- [x] outline2D_outer.gdshader
fixed with #70 
there is no shader error, although gdscript and scenes need update (mainly because Tween node was removed)
also I think you need to add .translation to gitignore, because those files are generated by godot

**Comment by NathanLovato** [comment]
Thanks, I'll do that

**Comment by NathanLovato** [comment]
Thanks, I'll do that
